### PR TITLE
Use linear slot identifiers within the wallet layer.

### PR DESCRIPTION
**Pull request by jonathanknowles**
# Issue Number

https://github.com/input-output-hk/cardano-wallet-adr/issues/10

# Overview

This change splits the existing `SlotId` type into two types:

1. A **linear slot identifier** type called `SlotId`. Values of this type are a simple count of the number of slots since the genesis slot (which has a value of 0).

    ```hs
    newtype SlotId = SlotId
        { getSlotId :: Word64
        } deriving stock (Bounded, Show, Read, Eq, Ord, Generic)
    ```

    This type is used throughout the wallet and DB layers. Slot arithmetic functions manipulate values of this type.

2. A **compound slot identifier** type called `EpochSlotId`:

    ```hs
    data EpochSlotId = EpochSlotId
        { epochNumber :: !Word64
        , slotNumber :: !Word16
        } deriving stock (Show, Read, Eq, Ord, Generic)
    ```
    This type is used _sparingly_, only in the API, and in places where we need to communicate across the network with back end nodes. The _network layer_ is responsible for converting values of this type to and from `SlotId`.

# Work Items

The following should be completed before this PR can be merged:

- [x] Provide support for `cardano-wallet-http-bridge`.
- [x] Figure out and fix cause of QSM test failures:
    * https://buildkite.com/input-output-hk/cardano-wallet/builds/2016#4a7468c8-f1ed-4fd4-b771-67090749b9fd/1406-3999
    * https://buildkite.com/input-output-hk/cardano-wallet/builds/2016#4a7468c8-f1ed-4fd4-b771-67090749b9fd/1406-5077

### CODE REVIEW

File: lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs [hunk 1]
```diff
@@ -905,30 +917,5 @@ instance Arbitrary EpochLength where
...
-    shrink (a,b) =
-        filter validSlotConfig $ zip (shrink a) (shrink b)
-      where
-        validSlotConfig (EpochLength ep, SlotId _ sl) = sl < ep
-
-    arbitrary = do
-        (EpochLength epochLength) <- arbitrary
-        ep <- choose (0, 1000)
-        sl <- choose (0, fromIntegral epochLength - 1)
-        return (EpochLength epochLength, SlotId ep sl)
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:941]: I particularly like the fact that we can now remove these overlapping instances.

**Comment by rvl** [comment]
I've squashed the commits, rebased on latest master, and attempted to fix the type skew. But there have been a lot of refactors in the short time since this PR was submitted, so it doesn't work yet.

**Comment by KtorZ** [comment]
We might want to push back a bit on this one until we have finalized the stake pool metrics. There, we need to know about epoch transitions (which could happen within the network layer though).

**Comment by jonathanknowles** [comment]
@KtorZ wrote:
>We might want to push back a bit on this one until we have finalized the stake pool metrics. There, we need to know about epoch transitions (which could happen within the network layer though).

Should we close this PR for now? We can always resurrect it later, when required.

**Pull request closed without merging**